src/encoded/audit/file.py

```diff
@@ -513,6 +513,53 @@ def audit_file_matching_md5sum(value, system):
     yield AuditFailure('matching md5 sums', detail, level='WARNING')
 
 
+def audit_file_index_of(value, system):
+    '''
+    Files with output_type: index reads may specify that they are index_of
+    either 1 SE or 2 PE files. The SE/PE files must be from the same
+    experiment as index, other combinations are flagged as ERROR.
+    https://encodedcc.atlassian.net/browse/ENCD-5252
+    '''
+    if value['output_type'] != 'index reads':
+        return
+    if 'index_of' in value:
+        index_exp = value['dataset']['@id']
+        fastq_ids = set()
+        exp_ids = set()
+        indexed_files = 0
+        error_files = 0
+        run_type = set()
+        for each in value['index_of']:
+            fastq_exp = each['dataset']
+            indexed_files += 1
+            run_type.add(each['run_type'])
+            if index_exp != fastq_exp:
+                fastq_ids.add(each['@id'])
+                exp_ids.add(each['dataset'])
+                error_files += 1
+        if error_files >= 1:
+            fastq_ids_links = [audit_link(path_to_text(m), m) for m in fastq_ids]
+            exp_ids_links = [audit_link(path_to_text(n), n) for n in exp_ids]
+            detail = ('Index file {} is from experiment {}, but is used as an index for file(s) {} from experiment(s) {}'.format(
+                                audit_link(path_to_text(value['@id']), value['@id']),
+                                audit_link(path_to_text(value['dataset']['@id']), value['dataset']['@id']),
+                                ', '.join(fastq_ids_links),
```

Comment:
I think we agreed to use f-strings for audit details and in general for strings in the code base. And I think that if you have multiple files that lead to multiple experiments, your details will be incorrect because file IDs in set will not necessarily be sorted like experiment IDs in the set.

Reply:
I reworked each detail into an f-string. I tried creating sorted tuples from a dict with the file/experiment pairs, and I think this is working to keep the order the same for each (I'm not sure how else to test it, but I triggered the audit with pairs of files several times in a local demo and they appeared in the correct order)

Reply:
I think because lists in python are ordered, if I append the new items to the end, the order should match in separate lists for files and fastqs- I returned to using two lists rather than dict used to then generate sorted lists.